fix devflags empty object

AjayJagan opened the pull request:
When an instance of DSC is created, it automatically adds `devFlags: {}` which needs to be only present if the manifests are provided.

## Description
issue: https://github.com/opendatahub-io/opendatahub-operator/issues/655

## How Has This Been Tested?
 Tested in local.

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

zdtsw commented:
can  you post the test result here?

AjayJagan commented:
> can you post the test result here?

Hi @zdtsw , I ran the `make test` and looks like there is an error. But I am not sure if this is related.
https://gist.github.com/AjayJagan/75b29ccbbcc410263e455796f1910769

zdtsw commented:
> > can you post the test result here?
> 
> Hi @zdtsw , I ran the `make test` and looks like there is an error. But I am not sure if this is related. https://gist.github.com/AjayJagan/75b29ccbbcc410263e455796f1910769

you will need to test it in the cluster with different combination, say, some set devflags, some not etc, to cover all case.
"make run" wont expose the behavior if regression is broken or not, e2e only has limited coverage and unit test is only for DSCI for the moment.

AjayJagan commented:
> > > can you post the test result here?
> > 
> > 
> > Hi @zdtsw , I ran the `make test` and looks like there is an error. But I am not sure if this is related. https://gist.github.com/AjayJagan/75b29ccbbcc410263e455796f1910769
> 
> you will need to test it in the cluster with different combination, say, some set devflags, some not etc, to cover all case. "make run" wont expose the behavior if regression is broken or not, e2e only has limited coverage and unit test is only for DSCI for the moment.

oh you mean that... I have tested that part and here is the yaml which is being generated
For this cr
```  apiVersion: datasciencecluster.opendatahub.io/v1
  kind: DataScienceCluster
  metadata:
    name: example
  spec:
    components:
      codeflare:
        managementState: Removed
      dashboard:
        managementState: Managed
        devFlags:
            manifests:
                - uri: https://github.com/opendatahub-io/odh-dashboard/tarball/incubation
      datasciencepipelines:
        managementState: Removed
      kserve:
        managementState: Removed
      modelmeshserving:
        managementState: Removed
      ray:
        managementState: Removed
      workbenches:
        devFlags: {}
        managementState: Managed```
        
I get this in the created yaml: https://gist.github.com/AjayJagan/75b29ccbbcc410263e455796f1910769

VaishnaviHire commented:
@AjayJagan can you rebase the PR, we can re-trigger the tests again

AjayJagan commented:
fixed issues.. let me know if this looks good @VaishnaviHire  @ykaliuta

ykaliuta commented on components/modelmeshserving/modelmeshserving.go:
```diff
@@ -33,31 +33,34 @@ type ModelMeshServing struct {
-			defaultKustomizePath := "base"
-			if subcomponent.SourcePath != "" {
-				defaultKustomizePath = subcomponent.SourcePath
+	if m.DevFlags != nil && len(m.DevFlags.Manifests) != 0 {
```
len() check is redundant here (the same in many other places, but it was there before the patch). I personally would make  if m.DevFlags == nil { return nil} to avoid this extra indentation change and make intentions clear, but looks like it does not fit the style.

AjayJagan replied:
Previously it was just `if len(m.DevFlags.Manifests) != 0 `
but with the addition of the `*DevFlags`, while the unmarshalling happened, it was a `nil` field and trying to access a `nil` field like `m.DevFlags.Manifests` creates a error(here DevFlags being `nil`). So I added the check.

AjayJagan replied:
removed the redundant len checks now and returned nil if DevFlags is nil early in the code :)

AjayJagan commented:
> Should the same be done for DSCI?

I am not sure if the DevFlags should still be there 😅 . I guess the idea was to have individual DevFlags instead of top level one, so maybe @zdtsw can you comment on this please

zdtsw commented on components/dashboard/dashboard.go:
```diff
@@ -48,6 +48,9 @@ type Dashboard struct {
 }
 
 func (d *Dashboard) OverrideManifests(platform string) error {
+	if d.DevFlags == nil {
+		return nil
+	}
```
why add the logic inside of the OverrideManifests() but not from the caller?  e.g in line 102?

AjayJagan replied:
hmm, so this logic cannot be added to the reconciler right? I mean it will return nil for reconciler and not proceed. I might have misunderstood your comment 😅  so please let me know how you want it

zdtsw replied:
maybe i was not clear at first:

```
if d.DevFlags != nil {
      if err = c.OverrideManifests(string(platform)); err != nil {
			return err
  }
}
```

AjayJagan replied:
oh that can be definitely done. should I change it everywhere?

AjayJagan replied:
updated accordingly now 👍

zdtsw commented on components/dashboard/dashboard.go:
```diff
@@ -127,7 +130,7 @@ func (d *Dashboard) ReconcileComponent(cli client.Client, owner metav1.Object, d
 
 		// Update image parameters (ODH does not use this solution, only downstream)
-		if dscispec.DevFlags.ManifestsUri == "" && len(d.DevFlags.Manifests) == 0 {
+		if d.DevFlags != nil && dscispec.DevFlags != nil && dscispec.DevFlags.ManifestsUri == "" && len(d.DevFlags.Manifests) == 0 {
```
this part i am not following:
1. DSCI devflags does not set ManifetsUri 
2. DSCI devFlags is set  
3. component devFlags is set
4. component devFlags does not set ManifestsUri
how could 1 and 2 happen at the same time  and 3 and 2 happen at the same time.
should that be just 
if d.DevFlags == nil && dscispec.DevFlags == nil  { // do it...}

AjayJagan replied:
since we have added a `*DevFlags` to the DSCI as well, it will be nil unless the user sets it in the yaml
So in this case dscispec.DevFlags.ManifestUri will result in error because DevFlags will be nil( it will be like dscispec.nil.ManifestsUri) so that is why I added the check. Or should we not add a pointer to the DevFlags in DSCI? wdyt

ykaliuta replied:
Hmm, does the original code mean "If neither dsci nor componet's manifests provided"? (not_dsci && not_component)

then one is not provided if the pointer is nil or the value is let's say blank.
not_dsci := (dscispec.DevFlags == nil || dscispec.DevFlags.ManifestsUri == "")
not_component := (d.DevFlags == nil || len(d.DevFlags.Manifests) == 0)

zdtsw replied:
let' break down it into 2 pieces, DSC and DSCI.
both are having the same logic in this if-condition.

what should trigger the caller of deploy.ApplyParams() is when user does not set DSC devFlags ManifestUri and user does not set component Devflags Manifests.

is this matching your implementation?

AjayJagan replied:
something like this then
`		if (dscispec.DevFlags == nil || dscispec.DevFlags.ManifestsUri == "") && (c.DevFlags == nil || len(c.DevFlags.Manifests) == 0) {
			if err := deploy.ApplyParams(CodeflarePath+"/bases", c.SetImageParamsMap(imageParamMap), true); err != nil {
				return err
			}
		}`

AjayJagan replied:
used this way now `(not_dsci && not_component)`

zdtsw reviewed (approved):
/lgtm
